src/commons/editor/useComments.tsx

```diff
@@ -0,0 +1,261 @@
+import { require as acequire, Ace } from 'ace-builds';
+import * as React from 'react';
+import ReactDOM from 'react-dom';
+
+import { groupBy, map, omit, sortBy, values, keyBy } from 'lodash';
+
+import { EditorHook } from './Editor';
+import Comments, { IComment } from './Comments';
+import { v1 as uuidv1 } from 'uuid';
+import { EventEmitter } from 'events';
+const LineWidgets = acequire('ace/line_widgets').LineWidgets;
+
+// Inferred from: https://github.com/ajaxorg/ace/blob/master/lib/ace/ext/error_marker.js#L129
+interface IWidget {
+  row: number;
+  fixedWidth: boolean;
+  coverGutter: boolean;
+  el: Element;
+  type: string;
+}
+
+interface ILineManager {
+  attach: (editor: Ace.Editor) => void;
+  addLineWidget: (widget: IWidget) => void;
+  removeLineWidget: (widget: IWidget) => void;
+}
+
+export interface CommentAPI {
+  // Assumes commentId is immutable. Do not update it unless explicitly via API.
+  updateComment: (...comments: IComment[]) => void;
+  updateCommentsBeingEdited: (...comments: IComment[]) => void;
+  removeComment: (comment: IComment) => void;
+  removeCommentEdit: (comment: IComment) => void;
+  // Probably need a replaceId for comments to update when the server does.
```

Comment:
I assume this is for the server to handle final serialisation of comment IDs right? Since it is possible for two users to have a pending comment with the same ID, so the server needs to be the coordinator of IDs?